From 8bae4419ce636f6f8414193a206a2dc2e6dd37db Mon Sep 17 00:00:00 2001
From: Namjae Jeon <namjae.jeon@samsung.com>
Date: Wed, 26 May 2021 15:24:39 +0900
Subject: [PATCH] cifsd: add goto fail in neg_token_init_mech_type()
Git-commit: 8bae4419ce636f6f8414193a206a2dc2e6dd37db
Patch-mainline: v5.15-rc1
References: bsn#12
Modified-by-SEL: No


Add goto fail in neg_token_init_mech_type().

Reviewed-by: Dan Carpenter <dan.carpenter@oracle.com>
Signed-off-by: Namjae Jeon <namjae.jeon@samsung.com>
Signed-off-by: Steve French <stfrench@microsoft.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 fs/cifsd/asn1.c | 19 ++++++++++---------
 1 file changed, 10 insertions(+), 9 deletions(-)

diff --git a/fs/cifsd/asn1.c b/fs/cifsd/asn1.c
index a7db37eef2a9..769fa328c020 100644
--- a/fs/cifsd/asn1.c
+++ b/fs/cifsd/asn1.c
@@ -284,14 +284,10 @@ int neg_token_init_mech_type(void *context, size_t hdrlen,
 	unsigned long *oid;
 	size_t oidlen;
 	int mech_type;
+	char buf[50];
 
-	if (!asn1_oid_decode(value, vlen, &oid, &oidlen)) {
-		char buf[50];
-
-		sprint_oid(value, vlen, buf, sizeof(buf));
-		ksmbd_debug(AUTH, "Unexpected OID: %s\n", buf);
-		return -EBADMSG;
-	}
+	if (!asn1_oid_decode(value, vlen, &oid, &oidlen))
+		goto fail;
 
 	if (oid_eq(oid, oidlen, NTLMSSP_OID, NTLMSSP_OID_LEN))
 		mech_type = KSMBD_AUTH_NTLMSSP;
@@ -302,15 +298,20 @@ int neg_token_init_mech_type(void *context, size_t hdrlen,
 	else if (oid_eq(oid, oidlen, KRB5U2U_OID, KRB5U2U_OID_LEN))
 		mech_type = KSMBD_AUTH_KRB5U2U;
 	else
-		goto out;
+		goto fail;
 
 	conn->auth_mechs |= mech_type;
 	if (conn->preferred_auth_mech == 0)
 		conn->preferred_auth_mech = mech_type;
 
-out:
 	kfree(oid);
 	return 0;
+
+fail:
+	kfree(oid);
+	sprint_oid(value, vlen, buf, sizeof(buf));
+	ksmbd_debug(AUTH, "Unexpected OID: %s\n", buf);
+	return -EBADMSG;
 }
 
 int neg_token_init_mech_token(void *context, size_t hdrlen,
-- 
2.33.0

